Assets/MixedRealityToolkit.SDK/Experimental/Features/UX/WindowsKeyboard/MixedRealityKeyboard.cs

```diff
@@ -0,0 +1,196 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+using UnityEngine;
+using UnityEngine.Events;
+
+#if !UNITY_EDITOR && UNITY_WSA
+using Windows.UI.ViewManagement;
+#endif 
+
+namespace Microsoft.MixedReality.Toolkit.Experimental.UI
+{
+    /// <summary>
+    /// Class that can launch and hide a system keyboard specifically for Windows Mixed Reality
+    /// devices (HoloLens 2, Windows Mixed Reality).
+    /// 
+    /// Implements a workaround for UWP TouchScreenKeyboard bug which prevents
+    /// UWP keyboard from showing up again after it is closed.
+    /// Unity bug tracking the issue https://fogbugz.unity3d.com/default.asp?1137074_rttdnt8t1lccmtd3
+    /// </summary>
+    public class MixedRealityKeyboard : MonoBehaviour
+    {
+        [Experimental]
+        #region Serialized fields
+        [SerializeField]
+        [Tooltip("Event raised when the text has changed")]
+        public KeyboardTextEvent TextChanged = new KeyboardTextEvent();
+        [Tooltip("Event raised when the keyboard is shown")]
+        public UnityEvent KeyboardShown = new UnityEvent();
+        [Tooltip("Event raised when the keyboard is hidden")]
+        public UnityEvent KeyboardHidden = new UnityEvent();
+        #endregion
+
+        #region Properties
+        private string text;
+        public string Text
+        {
+            get
+            {
+                return text;
+            }
+
+            private set
+            {
+                if (text != value)
+                {
+                    var oldValue = text;
+                    text = value;
+                    TextChanged?.Invoke(new KeyboardTextData(oldValue, value));
+                }
+            }
+        }
+        
+        private KeyboardState State
+        {
+            get
+            {
+                return state;
+            }
+
+            set
+            {
+                if (state != value)
+                {
+                    state = value;
+                    if (state == KeyboardState.showing)
+                    {
+                        KeyboardShown.Invoke();
+                    }
+                    else if (state == KeyboardState.hidden)
+                    {
+                        KeyboardHidden.Invoke();
+                    }
+                }
+            }
+        }
+        #endregion properties
+
+        #region Private fields
+        private TouchScreenKeyboard keyboard;
+
+#if !UNITY_EDITOR && UNITY_WSA
+    private InputPane inputPane = null;
+#endif //!UNITY_EDITOR && UNITY_WSA
+
+        private KeyboardState state = KeyboardState.hidden;
+        #endregion private fields
+
+        #region Private enums
+        private enum KeyboardState
+        {
+            hidden,
+            showing,
```

Comment:
>showing [](start = 12, length = 7)

Super nit: Order (hidding, hidden, showing (or showing, hiding, hidden). It's currently somewhat random